openpype/plugins/publish/integrate_version_attrs.py

```diff
@@ -0,0 +1,92 @@
+import pyblish.api
+import ayon_api
+
+from openpype import AYON_SERVER_ENABLED
+from openpype.client.operations import OperationsSession
+
+
+class IntegrateVersionAttributes(pyblish.api.ContextPlugin):
+    """Integrate version attributes from predefined key.
+
+    Any integration after 'IntegrateAsset' can fill 'versionAttributes' with
+    attribute key & value to be updated on created version.
+
+    The integration must make sure the attribute is available for the version
+    entity otherwise an error would be raised.
+
+    Example of 'versionAttributes':
+        {
+            "ftrack_id": "0123456789-101112-131415",
+            "syncsketch_id": "987654321-012345-678910"
+        }
+    """
+
+    label = "Integrate Version Attributes"
+    order = pyblish.api.IntegratorOrder + 0.5
+
+    def process(self, context):
+        available_attributes = ayon_api.get_attributes_for_type("version")
+        skipped_attributes = set()
+        project_name = context.data["projectName"]
+        op_session = OperationsSession()
+        for instance in context:
+            label = self.get_instance_label(instance)
+            version_entity = instance.data.get("versionEntity")
+            if not version_entity:
+                continue
+            attributes = instance.data.get("versionAttributes")
+            if not attributes:
+                self.log.debug((
+                    "Skipping instance {} does not have set attributes"
+                ).format(label))
+                continue
+
+            filtered_attributes = {}
+            for attr, value in attributes.items():
+                if attr not in available_attributes:
+                    skipped_attributes.add(attr)
+                else:
+                    filtered_attributes[attr] = value
+
+            if not filtered_attributes:
+                self.log.debug((
+                    "Skipping instance {} all attributes were filtered"
+                ).format(label))
+                continue
+
+            self.log.debug("Updated attributes on version {} to {}".format(
+                version_entity["_id"], str(filtered_attributes)
+            ))
+            op_session.update_entity(
+                project_name,
+                "version",
+                version_entity["_id"],
+                {"attrib": filtered_attributes}
+            )
+
+        if skipped_attributes:
+            self.log.warning(
+                "Attributes that are not available on a server {}".format(
+                    str(skipped_attributes)
+                )
+            )
```

Comment:
I feel like this warning message can be clearer. _"On a server"_ what server? Did you mean to write Ayon server? "Attributes", what attributes or for what?

I imagine something like this might read clearer when just reading the logs: `"Attributes skipped due to not being available attributes for version entities on the Ayon server: {}"`

Reply:
> Also, any reason we're logging this once for the full context instead of explicitly per instance?

For me it is the speed (single server call instead of one per instance). But I can change it to instance plugin if it's a big issue...

Reply:
I didn't mean the actual committing of data, but I meant the logging of skipped attributes.

Reply:
Aaah, I don't think it would give you more information. It is expected that there will be 1-3 attributes (max) that could be set during publishing and they would be the same for each instance -> report of the same information multiple times is useless.

Reply:
> I imagine something like this might read clearer when just reading the logs: "Attributes skipped due to not being available attributes for version entities on the Ayon server: {}"

I had to read it twice (and I know what it should tell) :D